concurrent/src/main/java/io/airlift/concurrent/ResourceLoader.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.File;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public final class ResourceLoader<T>
+{
+    private static final Logger log = Logger.get(ResourceLoader.class);
+
+    private final File file;
```

Comment:
Maybe generify `File` into S (source). This could be Path, URL or id of an auxiliary resource.